### PR TITLE
Optimize 'GetSystemLockdownPolicy' for non-lockdown scenarios

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Today, the implementation of `GetSystemLockdownPolicy` is optimized for lockdown scenarios, but dis-optimized for the more common non-lockdown scenario because the test hook will make this method call the Win32 API `WldpGetLockdownPolicy` every time. This PR attempts to improve it.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@PaulHigin You mentioned the registry check in [GetDebugLockdownPolicy](https://github.com/PowerShell/PowerShell/blob/d9c2e0e712361bd91f661f6bf6a010c01d2eac0a/src/System.Management.Automation/security/wldpNativeMethods.cs#L360-L384) may not be needed anymore for our tests, can you please review and confirm if that's the case? If so, I can remove that part in this PR.

File: src/System.Management.Automation/security/wldpNativeMethods.cs [hunk 1]
```diff
@@ -48,16 +48,23 @@ private SystemPolicy()
...
-            if (s_allowDebugOverridePolicy || (s_systemLockdownPolicy == null))
+            if (s_systemLockdownPolicy == null)
             {
                 lock (s_systemLockdownPolicyLock)
                 {
-                    if (s_allowDebugOverridePolicy || (s_systemLockdownPolicy == null))
+                    if (s_systemLockdownPolicy == null)
                     {
-                        s_systemLockdownPolicy = GetLockdownPolicy(null, null);
+                        s_systemLockdownPolicy = GetLockdownPolicy(path: null, handle: null);
```
Comment (TravisEz13) [inline src/System.Management.Automation/security/wldpNativeMethods.cs:57]: First call here, we still make the win32 call.

Comment (TravisEz13) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:57]: if there is no policy, `s_allowDebugOverridePolicy` is true and `s_systemLockdownPolicy` is none.

File: src/System.Management.Automation/security/wldpNativeMethods.cs [hunk 2]
```diff
@@ -48,16 +48,23 @@ private SystemPolicy()
...
                     }
                 }
             }
+            else if (s_allowDebugOverridePolicy)
```
Comment (TravisEz13) [inline src/System.Management.Automation/security/wldpNativeMethods.cs:61]: So here, on subsequent calls we only make the call for the hook?

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 1 above)
Comment (TravisEz13) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:57]: continued in next comment.

Comment (daxian-dbw) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:57]: > First call here, we still make the win32 call.

Yes, I believe that's needed, right?

File: src/System.Management.Automation/security/wldpNativeMethods.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:61]: Yes. If `s_allowDebugOverridePolicy` is true, that means the first call to `GetLockdownPolicy` did make into `GetDebugLockdownPolicy` without returning earlier, namely there is no platform policy enforced. So subsequent calls will just exercise the test hook code, without having to call the `Wldp` and `AppLocker` apis again.

Comment (TravisEz13) [reply src/System.Management.Automation/security/wldpNativeMethods.cs:61]: Yup, I was making sure I understood the change.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
These changes LGTM.  However, please also remove the debug registry check.  AFAIK this is no longer needed and tests use the `__PSLockdownPolicy` environment variable. and 'System32' filename append.  If any tests fail, we can look at them.

https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/security/wldpNativeMethods.cs#L360

I want to remove the registry check because users sometimes believe it is a 'secure' way to lock down PowerShell, which it is not.  Since we are PowerShell Core I feel we can remove it safely.  We can always re-add it if needed.

**Comment by daxian-dbw** [comment]
@TravisEz13 and @PaulHigin, thanks for the review!
I have removed the registry check in the debug hook code as @PaulHigin suggested.

**Review by PaulHigin — APPROVED** [review]
Thanks for making the change!

**Comment by daxian-dbw** [comment]
@TravisEz13 I think this PR is ready to merge.

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**